pkg/runtime/runtime.go

```diff
@@ -655,21 +654,23 @@ func (a *DaprRuntime) subscribeTopic(parentCtx context.Context, name string, top
 	policy := a.resiliency.ComponentInboundPolicy(ctx, name)
 	err := a.pubSubs[name].component.Subscribe(ctx, pubsub.SubscribeRequest{
 		Topic:    topic,
-		Metadata: route.metadata,
+		Metadata: route.routes.metadata,
 	}, func(ctx context.Context, msg *pubsub.NewMessage) error {
 		if msg.Metadata == nil {
 			msg.Metadata = make(map[string]string, 1)
 		}
 
 		msg.Metadata[pubsubName] = name
 
-		rawPayload, err := contrib_metadata.IsRawPayload(route.metadata)
+		rawPayload, err := contrib_metadata.IsRawPayload(route.routes.metadata)
 		if err != nil {
 			log.Errorf("error deserializing pubsub metadata: %s", err)
-			if configured, dlqErr := a.sendToDeadLetterIfConfigured(name, msg); configured && dlqErr == nil {
-				// dlq has been configured and message is successfully sent to dlq.
-				diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
-				return nil
+			if route.deadLetterTopic != "" {
+				if dlqErr := a.sendToDeadLetter(name, msg, route.deadLetterTopic); dlqErr == nil {
+					// dlq has been configured and message is successfully sent to dlq.
+					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
+					return nil
+				}
```

Comment:
If I'm not wrong, this code is pretty much the same as lines 686, 700, and 725,

do you think it worth extracting it to an internal func and reusing it instead? it could be a long-named function like

sendDeadLetter[andPublishIngressEvent]

Reply:
Sadly we cannot do that. We could if Go supported macros...

The reason is that in line 672, we want to return from the entire function, but ONLY if sending to the DLQ did not error. Otherwise, we want to continue execution and hit the return on line 676

Reply:
how about having a flag for it? you think it's a good idea?

like

if ok := a.trySendToDeadLetter(...); ok {
    return nil
}